tree/ntuple/v7/doc/architecture.md

```diff
@@ -3,7 +3,439 @@ RNTuple Code Architecture
+The RNTuple class contains the information of the RNTuple anchor in a ROOT file (see specification).
+It has a dictionary and is streamed through the standard ROOT I/O.
+An RNTuple object represents an RNTuple dataset but it is not the dataset itself.
+It can be used like a token to open the actual RNTuple dataset with, e.g., RDF or an RNTupleReader,
+and it provides the `Merge(...)` interface for the `TFileMerger`.
+
+### RPageSource / Sink
+The page source and sink can read and write pages and clusters from and to a storage backend.
+There are concrete class implementations for an RNTuple stored in a ROOT file (local or remote), and for an RNTuple stored in a DAOS object store.
+There is a virtual page sink for buffered writes, which also groups pages of the same column before flushing them to disk.
+There is a virtual page source for aligned friend datasets (horizontal data combination).
+
+Page sources and sinks do not operate entry-based but based on pages/indices of columns.
+For instance, there is no API in the page sink to write an entry, but only to write pages of columns.
+The higher-level APIs, e.g. `RField`, `REntry`, `RNTupleWriter`, take care of presenting the available data as entries where necessary.
+
+The page source also gives access to an `RNTupleDescriptor` through a read/write lock guard.
+The `RNTupleDescriptor` owned by the page source changes only when new cluster meta-data are loaded.
+The header and the cluster group summary information is stable throughout its lifetime (cf. format specification).
+
+### R{NTuple,Field,Column,Cluster,...}Descriptor
+The descriptor classes provide read-only access to the on-disk meta-data of an RNTuple.
+The meta-data include the schema (fields and columns), information about clusters and the page locations.
+The descriptor classes are closely related to the format specification.
+
+For normal read and write tasks, access to the descriptor is not necessary.
+One notable exception is bulk reading, where the descriptor can be used to determine entry boundaries of clusters.
+The descriptors are used internally, e.g. to build an RNTupleModel from the on-disk information.
+The descriptors are also useful for inspection purposes.
+
+The descriptor classes contain a copy of the meta-data; they are not linked to an open page source.
+A descriptor can be used after its originating page source has been deleted.
+
+### RField<T>
+The RField<T> classes are central in RNTuple:
+they link the in-memory representation of data types to their on-disk representation.
+All field classes inherit from `RFieldBase`.
+
+Every type with RNTuple I/O supported has a corresponding RField<T> template specialization.
+Complex types are composed of fields and sub fields.
+E.g., a `struct` is represented by a parent field for the `struct` itself and a subfield for every member of the `struct`.
+Fields of complex types have type-erased versions in addition to the templated ones (e.g., `RVectorField`, `RClassField`).
```

Comment:
```suggestion
Fields of complex types have type-erased versions in addition to the templated ones (e.g., `RCollectionField`, `RClassField`).
```

Reply:
I think here I would like to stick to the `RVectorField`. The `RCollectionField` is special in that it makes the connection between an RNTupleCollectionWriter and the RNTupleWriter.